src/imp/span_imp.js

```diff
@@ -1,119 +1,122 @@
-        }
+    _addTags(keyValuePairs) {
+        let self = this;
+        _each(keyValuePairs, (value, key) => {
+            self._tags[key] = value;
+        });
     }
 
-    /**
-     * The set of OpenTracing fields is:
-     * - 'event'
-     * - 'timestamp'
-     * - 'payload'
-     */
-    // Log record specified by fields
-    log(fields) {
-        const argumentType = typeof fields;
+    _log(keyValuePairs, timestamp) {
+        let self = this;
+        const argumentType = typeof keyValuePairs;
         if (argumentType !== 'object') {
-            this._tracer._error('Span.log() expects an object as its argument');
+            self._tracerImp._error('Span.log() expects an object as its first argument');
             return;
         }
 
-        let rec = this._tracer.log()
-            .span(this.guid())
-            .level(constants.LOG_STRING_TO_LEVEL[fields.level] || constants.LOG_INFO);
-
-        //
-        // OpenTracing attributes
-        //
-        if (fields.event !== undefined) {
-            rec.name(fields.event);
-        }
-        if (fields.timestamp !== undefined) {
-            // The incoming 'timestamp' field is in milliseconds. The internal
-            // timestamp is in microseconds.
-            rec.timestamp(fields.timestamp * 1000);
-        }
-        if (fields.payload !== undefined) {
-            rec.payload(fields.payload);
-        }
-
-        //
-        // LightStep-specific attributes
-        //
-        if (fields.message !== undefined) {
-            rec.message(fields.message);
-        }
-        if (fields.timestamp_micros !== undefined) {
-            rec.timestamp(fields.timestamp_micros);
+        let tsMicros = null;
+        if (timestamp) {
+            tsMicros = timestamp * 1000;
+        } else {
+            tsMicros = self._tracerImp._platform.nowMicros();
         }
-        rec.end();
+        let fields = [];
+        _each(keyValuePairs, (value, key) => {
+            if (!key || !value) {
+                return;
+            }
+            let keyStr = coerce.toString(key);
+            let valStr = null;
+            if (value instanceof Object) {
+                valStr = JSON.stringify(value, null, "  ")
+            } else {
+                valStr = coerce.toString(value);
+            }
+            if (keyStr.length > self._tracerImp._options.log_field_key_hard_limit) {
+                self._tracerImp._counters['logs.keys.over_limit']++;
+                keyStr = keyStr.substr(0, self._tracerImp._options.log_field_key_hard_limit) + "...";
+            }
+            if (valStr.length > self._tracerImp._options.log_field_value_hard_limit) {
+                self._tracerImp._counters['logs.values.over_limit']++;
+                valStr = valStr.substr(0, self._tracerImp._options.log_field_value_hard_limit) + "...";
+            }
+            fields.push(new crouton_thrift.KeyValue({
+                Key   : keyStr,
+                Value : valStr,
```

Comment:
Is `null` a valid value for `Value`? Or should we skip the `push` entirely? I don't recall off the top of my head how Thrift and the collector process a value of `null`.